src/LeanCode.ContractsGenerator/Generation/TypeRefFactory.cs

```diff
@@ -62,6 +62,56 @@ public TypeRef From(ITypeSymbol symbol)
         }
     }
 
+    public NotificationTypeRef FromNotification(ITypeSymbol symbol)
+    {
+        return new()
+        {
+            Type = From(symbol),
+            Tag = GetFullTypeName(symbol),
+        };
+    }
+
+    private static string GetFullTypeName(ITypeSymbol typeSymbol)
+    {
+        // Represents a type other than an array, a pointer, a type parameter.
+        if (typeSymbol is INamedTypeSymbol namedTypeSymbol)
+        {
+            var typeName = namedTypeSymbol.ContainingNamespace.IsGlobalNamespace
+                ? namedTypeSymbol.Name
+                : $"{namedTypeSymbol.ContainingNamespace}.{namedTypeSymbol.Name}";
+
+            if (namedTypeSymbol.TypeArguments.Length > 0)
+            {
+                typeName += "`" + namedTypeSymbol.TypeArguments.Length + "[";
+
+                for (var i = 0; i < namedTypeSymbol.TypeArguments.Length; i++)
+                {
+                    if (i > 0)
+                    {
+                        typeName += ",";
+                    }
+
+                    var typeArg = GetFullTypeName(namedTypeSymbol.TypeArguments[i]);
+                    typeName += typeArg;
+                }
+
+                typeName += "]";
+            }
+
+            return typeName;
+        }
+        else if (typeSymbol is IArrayTypeSymbol arrayTypeSymbol)
+        {
+            var elementType = GetFullTypeName(arrayTypeSymbol.ElementType);
+            var rank = arrayTypeSymbol.Rank;
+            var dimensions = new string(',', rank - 1);
+
+            return elementType + "[" + dimensions + "]";
+        }
```

Comment:
This makes me wonder why we even support all those list-like containers (`IList`, `List`, `T[]`, `IReadOnlyList`, etc). In my opinion sticking to only one would be perfectly acceptable (and preferred) for contracts. 

This generates various tags for things that are seen as equivalent by the semantic model of the contracts. For instance:

```cs
class Auction : ITopic, IProduceNotification<int[]>, IProduceNotification<List<int>> {}
```

Both notifications have the same `TypeRef`. 

---

This actually reminds me of a different issue. I wonder how clients are supposed to name different variants of notifications in a topic. On Dart's side when subscribing to a topic we cannot say that it will emit notifications of a top type (`dynamic` in Dart), otherwise we don't benefit at all from contracts being typed. So instead we allow switching over all possible notification types. But this in one way or the other requires naming individual notification types. If we have something like `IProduceNotification<int[]>` I don't see any good way to name such a notification type. Typescript will probably be able to get away with some typing using the string tag itself.

Reply:
> This makes me wonder why we even support all those list-like containers (IList, List, T[], IReadOnlyList, etc). In my opinion sticking to only one would be perfectly acceptable (and preferred) for contracts.

I don't know why other array types are allowed as well, but @jakubfijalkowski or @Saancreed should have an opinion on that.

Reply:
> If we have something like IProduceNotification<int[]> I don't see any good way to name such a notification type.

While I agree such type has little semantics, it may have sense in some very simple, non-context driven. For naming things, why not using the contracts generated `TypeTag`?

Reply:
> why not using the contracts generated `TypeTag`

The tag is not a valid identifier name, which it has to be to be usable on the type level in Dart. For TypeScript this won't be a problem. Unless I am missing something and there is a way to use it here.

Reply:
> This makes me wonder why we even support all those list-like containers (IList, List, T[], IReadOnlyList, etc). In my opinion sticking to only one would be perfectly acceptable (and preferred) for contracts.

Because sometimes arrays are required on .NET side. Being able to use them makes sense from performance reasons (you can't cast `List` to array, you have to transform it) _in other places of the contracts_.

Here, I am more and more inclined to allow only DTOs, so `TypeRef.Internal`. Unnamed (a generic name also counts) notifications don't make any sense from business perspective.